Implement in-place interpreter prologue and loop OSR

danlliu opened the pull request:
#### 5f18ff5fc08d864caa2315af403ec63f535362fa
<pre>
Implement in-place interpreter prologue and loop OSR
<a href="https://bugs.webkit.org/show_bug.cgi?id=260083">https://bugs.webkit.org/show_bug.cgi?id=260083</a>
rdar://113768807

Reviewed by Justin Michaud.

Add on stack replacement (OSR) functionality to the in-place interpreter, allowing tiering up to both BBQ and OMG JIT tiers. JS2 performance is comparable to LLInt, with local scores both at around 330.

* JSTests/wasm/ipint-tests/ipint-test-hot-12.js: Added.
(prng):
(async test):
* JSTests/wasm/ipint-tests/ipint-test-hot-loop.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.i32.result.i32.local.i32.i32.local.0.i32.const.2.local.tee.1.i32.lt_s.then.i32.const.0.return.loop.local.0.local.1.i32.rem_s.then.local.1.i32.const.1.i32.add.local.1.else.i32.const.0.return.local.1.local.0.i32.lt_s.br_if.0.local.0.local.1.i32.eq.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-hot.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.i32.result.i32.local.i32.i32.i32.const.1.local.1.loop.local.2.local.1.local.2.i32.add.local.2.local.1.local.0.i32.const.1.i32.sub.local.tee.0.i32.const.0.i32.gt_s.br_if.0.local.2.return.async test):
* Source/JavaScriptCore/b3/B3Procedure.h:
(JSC::B3::Procedure::usesSIMD const):
* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter.cpp:
(ipint_probe_precall):
(ipint_probe_calling):
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::work):
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::IPIntCallee::IPIntCallee):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmCalleeGroup.h:
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
(JSC::Wasm::FunctionIPIntMetadataGenerator::tierUpCounter):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::notifyFunctionUsesSIMD):
(JSC::Wasm::IPIntGenerator::condenseControlFlowInstructions):
(JSC::Wasm::IPIntGenerator::addLoop):
(JSC::Wasm::IPIntGenerator::endTopLevel):
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
(JSC::Wasm::ModuleInformation::usesSIMD const):
* Source/JavaScriptCore/wasm/WasmOMGPlan.cpp:
(JSC::Wasm::OMGPlan::work):
* Source/JavaScriptCore/wasm/WasmOSREntryPlan.cpp:
(JSC::Wasm::OSREntryPlan::work):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::shouldJIT):
(JSC::LLInt::jitCompileAndSetHeuristics):
(JSC::LLInt::WASM_IPINT_EXTERN_CPP_DECL):
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:

Canonical link: <a href="https://commits.webkit.org/266887@main">https://commits.webkit.org/266887@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af4340a0691987da2f976dd13f806e4b95520778

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14918 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15222 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15576 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16666 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14033 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17743 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15323 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16667 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15098 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15578 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12671 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17399 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12857 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13457 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20423 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/12766 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13936 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13625 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16862 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/14180 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14177 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11986 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/15116 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13461 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3855 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3624 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17793 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/15347 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14019 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3662 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -986,21 +988,46 @@ PartialResult WARN_UNUSED_RETURN IPIntGenerator::addBlock(BlockSignature signatu
+
+    // We don't care what we're putting in as long as we put in something to represent a value
+    // Will get everything off the stack
+    Vector<VirtualRegister> osrEntryData;
```
Can we just initialize this vector directly to be the right size?

danlliu replied:
Updated to precalculate the size, and then initialize the vector directly with that size.

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
```diff
@@ -384,6 +384,92 @@ macro ipintException(exception)
+    push a0, a1, a2, a3, a4, a5, a6, a7
+    subq 32, sp
+    storepaird fa0, fa1, [sp]
+    storepaird fa2, fa3, 16[sp]
```
Is `fa3` enough? I think wasm ARM64 has upto wfa7.
It would be nice if we can use `forEachArgumentFPR` here.

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
```diff
@@ -384,6 +384,92 @@ macro ipintException(exception)
+    baddis increment, Wasm::IPIntCallee::m_tierUpCounter + Wasm::LLIntTierUpCounter::m_counter[ws0], .continue
+
+if ARM64 or ARM64E
+    push a0, a1, a2, a3, a4, a5, a6, a7
```
How about using `forEachArgumentJSR`?

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
```diff
@@ -384,6 +384,92 @@ macro ipintException(exception)
+    pop boundsCheckingSize, memoryBase
+if ARM64 or ARM64E
+    loadpaird [sp], fa0, fa1
+    loadpaird 16[sp], fa2, fa3
```
Ditto.

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
```diff
@@ -384,6 +384,92 @@ macro ipintException(exception)
+    loadpaird [sp], fa0, fa1
+    loadpaird 16[sp], fa2, fa3
+    addq 32, sp
+    pop a7, a6, a5, a4, a3, a2, a1, a0
```
Ditto.

justinmichaud reviewed (approved):
r=me

danlliu committed:
Implement in-place interpreter prologue and loop OSR
https://bugs.webkit.org/show_bug.cgi?id=260083
rdar://113768807

Reviewed by Justin Michaud.

Add on stack replacement (OSR) functionality to the in-place interpreter, allowing tiering up to both BBQ and OMG JIT tiers. JS2 performance is comparable to LLInt, with local scores both at around 330.

* JSTests/wasm/ipint-tests/ipint-test-hot-12.js: Added.
(prng):
(async test):
* JSTests/wasm/ipint-tests/ipint-test-hot-loop.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.i32.result.i32.local.i32.i32.local.0.i32.const.2.local.tee.1.i32.lt_s.then.i32.const.0.return.loop.local.0.local.1.i32.rem_s.then.local.1.i32.const.1.i32.add.local.1.else.i32.const.0.return.local.1.local.0.i32.lt_s.br_if.0.local.0.local.1.i32.eq.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-hot.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.i32.result.i32.local.i32.i32.i32.const.1.local.1.loop.local.2.local.1.local.2.i32.add.local.2.local.1.local.0.i32.const.1.i32.sub.local.tee.0.i32.const.0.i32.gt_s.br_if.0.local.2.return.async test):
* Source/JavaScriptCore/b3/B3Procedure.h:
(JSC::B3::Procedure::usesSIMD const):
* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter.cpp:
(ipint_probe_precall):
(ipint_probe_calling):
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::work):
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::IPIntCallee::IPIntCallee):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmCalleeGroup.h:
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
(JSC::Wasm::FunctionIPIntMetadataGenerator::tierUpCounter):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::notifyFunctionUsesSIMD):
(JSC::Wasm::IPIntGenerator::condenseControlFlowInstructions):
(JSC::Wasm::IPIntGenerator::addLoop):
(JSC::Wasm::IPIntGenerator::endTopLevel):
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
(JSC::Wasm::ModuleInformation::usesSIMD const):
* Source/JavaScriptCore/wasm/WasmOMGPlan.cpp:
(JSC::Wasm::OMGPlan::work):
* Source/JavaScriptCore/wasm/WasmOSREntryPlan.cpp:
(JSC::Wasm::OSREntryPlan::work):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::shouldJIT):
(JSC::LLInt::jitCompileAndSetHeuristics):
(JSC::LLInt::WASM_IPINT_EXTERN_CPP_DECL):
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:

Canonical link: https://commits.webkit.org/266887@main
